File: codes/python/ch07/cheb_convergence.py
#!/usr/bin/env python3
"""
cheb_convergence.py

Demonstrates spectral convergence rates for four functions of increasing
smoothness. Shows the fundamental relationship between function regularity
and convergence rate of spectral differentiation.

Test functions (different from Trefethen's standard examples):
1. |x|^(5/2) - Third derivative of bounded variation (algebraic: ~N^{-2.5})
2. exp(-1/(1-x^2)) - Smooth bump function, C^∞ but not analytic (superalgebraic)
3. tanh(5x) - Analytic in [-1,1], poles at ±iπ/10 (exponential)
4. x^8 - Polynomial of degree 8 (exact for N ≥ 8)

This script generates Figure 6.5 and Table 6.2 for Chapter 6.

Author: Dr. Denys Dutykh
        Mathematics Department
        Khalifa University of Science and Technology
        Abu Dhabi, UAE

Part of the book "Computational Etudes: A Spectral Approach"
https://github.com/dutykh/computational-etudes
"""

import numpy as np
import matplotlib.pyplot as plt
from matplotlib import rcParams
from pathlib import Path

# Import Chebyshev matrix function
import sys
sys.path.insert(0, str(Path(__file__).parent))
from cheb_matrix import cheb_matrix

# -----------------------------------------------------------------------------
# Publication-quality matplotlib configuration
# -----------------------------------------------------------------------------
rcParams['font.family'] = 'serif'
rcParams['font.serif'] = ['Computer Modern Roman', 'CMU Serif', 'DejaVu Serif']
rcParams['mathtext.fontset'] = 'cm'
rcParams['axes.labelsize'] = 11
rcParams['axes.titlesize'] = 12
rcParams['xtick.labelsize'] = 10
rcParams['ytick.labelsize'] = 10
rcParams['legend.fontsize'] = 10
rcParams['figure.dpi'] = 150
rcParams['savefig.dpi'] = 300
rcParams['text.usetex'] = False
rcParams['axes.linewidth'] = 0.8
rcParams['xtick.major.width'] = 0.8
rcParams['ytick.major.width'] = 0.8

# -----------------------------------------------------------------------------
# Book color scheme
# -----------------------------------------------------------------------------
NAVY = '#142D6E'
SKY = '#7896D2'
CORAL = '#E74C3C'
TEAL = '#16A085'
PURPLE = '#8E44AD'
ORANGE = '#E67E22'

# Output paths
SCRIPT_DIR = Path(__file__).parent
OUTPUT_DIR = SCRIPT_DIR / '..' / '..' / '..' / 'textbook' / 'figures' / 'ch06' / 'python'


# -----------------------------------------------------------------------------
# Test functions and their exact derivatives
# -----------------------------------------------------------------------------

# Function 1: |x|^(5/2) - C^2, 3rd derivative in BV
def f1(x):
    return np.abs(x) ** 2.5

def f1_prime(x):
    # d/dx |x|^(5/2) = (5/2)|x|^(3/2) * sign(x)
    return 2.5 * np.abs(x) ** 1.5 * np.sign(x)


# Function 2: exp(-1/(1-x^2)) - Smooth bump, C^∞ but not analytic at boundaries
def f2(x):
    result = np.zeros_like(x)
    mask = np.abs(x) < 1
    result[mask] = np.exp(-1.0 / (1.0 - x[mask]**2))
    return result

def f2_prime(x):
    result = np.zeros_like(x)
    mask = np.abs(x) < 1
    inner = 1.0 - x[mask]**2
    result[mask] = np.exp(-1.0 / inner) * (-2.0 * x[mask]) / inner**2
    return result


# Function 3: tanh(5x) - Analytic, poles at x = ±iπ/10
def f3(x):
    return np.tanh(5.0 * x)

def f3_prime(x):
    return 5.0 / np.cosh(5.0 * x)**2


# Function 4: x^8 - Polynomial degree 8
def f4(x):
    return x ** 8

def f4_prime(x):
    return 8.0 * x ** 7


# List of test functions
TEST_FUNCTIONS = [
    {'name': r'$|x|^{5/2}$', 'f': f1, 'fp': f1_prime,
     'smoothness': r'$C^2$, 3rd deriv in BV', 'rate': r'$O(N^{-2.5})$',
     'color': CORAL},
    {'name': r'$e^{-1/(1-x^2)}$', 'f': f2, 'fp': f2_prime,
     'smoothness': r'$C^\infty$, not analytic', 'rate': r'faster than any $N^{-k}$',
     'color': TEAL},
    {'name': r'$\tanh(5x)$', 'f': f3, 'fp': f3_prime,
     'smoothness': 'Analytic', 'rate': r'$O(\rho^{-N})$',
     'color': PURPLE},
    {'name': r'$x^8$', 'f': f4, 'fp': f4_prime,
     'smoothness': 'Polynomial', 'rate': r'Exact for $N \geq 8$',
     'color': ORANGE},
]


def main():
    """Create convergence waterfall figure."""

    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

    # Range of N values
    N_values = np.array([4, 6, 8, 10, 12, 16, 20, 24, 32, 40, 48, 64, 80, 96, 128])

    # Create 2x2 figure
    fig, axes = plt.subplots(2, 2, figsize=(10, 8))

    # Store errors for table
    error_data = {tf['name']: [] for tf in TEST_FUNCTIONS}

    for idx, (ax, tf) in enumerate(zip(axes.flat, TEST_FUNCTIONS)):
        errors = []

        for N in N_values:
            D, x = cheb_matrix(N)
            v = tf['f'](x)
            w = D @ v
            w_exact = tf['fp'](x)

            # Compute max error (avoiding boundary issues for bump function)
            if idx == 1:  # Bump function
                interior = (x > -0.99) & (x < 0.99)
                if np.any(interior):
                    error = np.max(np.abs(w[interior] - w_exact[interior]))
                else:
                    error = np.max(np.abs(w - w_exact))
            else:
                error = np.max(np.abs(w - w_exact))

            errors.append(error if error > 1e-16 else 1e-16)
            error_data[tf['name']].append((N, error))

        errors = np.array(errors)

        # Plot convergence
        ax.semilogy(N_values, errors, 'o-', color=tf['color'], linewidth=1.5,
                    markersize=5, markeredgecolor='white', markeredgewidth=0.5)

        # Add reference lines for theoretical rates
        if idx == 0:  # |x|^(5/2) - algebraic O(N^-2.5)
            N_ref = N_values[N_values >= 16]
            err_ref = errors[N_values >= 16]
            if len(err_ref) > 0:
                C = err_ref[0] * N_ref[0]**2.5
                ax.semilogy(N_ref, C * N_ref**(-2.5), '--', color='gray',
                            linewidth=1, alpha=0.7, label=r'$O(N^{-2.5})$')
        elif idx == 2:  # tanh(5x) - exponential
            # rho = distance to nearest singularity = pi/10 ≈ 0.314
            # So convergence rate ~ (1 + pi/10)^(-N) is too slow
            # Actually for tanh, poles at ±iπ/10 give exponential convergence
            pass
        elif idx == 3:  # x^8 - exact for N >= 8
            ax.axhline(1e-14, color='gray', linewidth=1, linestyle='--', alpha=0.7)

        ax.set_xlabel(r'$N$')
        ax.set_ylabel('Max Error')
        ax.set_title(f'{tf["name"]}: {tf["smoothness"]}', fontsize=11)
        ax.set_xlim(0, 135)
        ax.set_ylim(1e-16, 1e2)
        ax.grid(True, which='both', alpha=0.3)
        ax.spines['top'].set_visible(False)
        ax.spines['right'].set_visible(False)

        # Add convergence rate annotation
        ax.text(0.95, 0.95, tf['rate'], transform=ax.transAxes,
                fontsize=9, verticalalignment='top', horizontalalignment='right',
                bbox=dict(boxstyle='round', facecolor='white', alpha=0.8))

    # Main title
    fig.suptitle('Spectral Convergence: Four Functions of Increasing Smoothness',
                 fontsize=13, y=0.98)

    plt.tight_layout(rect=[0, 0, 1, 0.96])

    # Save figure
    output_file = OUTPUT_DIR / 'convergence_waterfall.pdf'
    fig.savefig(output_file, bbox_inches='tight', pad_inches=0.05)
    fig.savefig(output_file.with_suffix('.png'), bbox_inches='tight',
                pad_inches=0.05, dpi=300)

    print(f'Figure saved to: {output_file.resolve()}')
    plt.close(fig)

    # Generate convergence table
    print("\n" + "=" * 80)
    print("Convergence Table: Chebyshev Spectral Differentiation")
    print("=" * 80)

    # Table header
    print(f"{'N':>6}", end='')
    for tf in TEST_FUNCTIONS:
        name_short = tf['name'].replace('$', '').replace('\\', '').replace('{', '').replace('}', '')
        print(f"  {name_short:>14}", end='')
    print()
    print("-" * 80)

    # Table rows
    for i, N in enumerate(N_values):
        print(f"{N:>6d}", end='')
        for tf in TEST_FUNCTIONS:
            _, err = error_data[tf['name']][i]
            if err < 1e-14:
                print(f"  {'< 1e-14':>14}", end='')
            else:
                print(f"  {err:>14.2e}", end='')
        print()

    print("=" * 80)

    # Summary
    print("\nSummary:")
    print("-" * 80)
    print(f"{'Function':<25} {'Smoothness':<25} {'Expected Rate':<20}")
    print("-" * 80)
    for tf in TEST_FUNCTIONS:
        name_clean = tf['name'].replace('$', '')
        smooth_clean = tf['smoothness'].replace('$', '').replace('\\', '')
        rate_clean = tf['rate'].replace('$', '').replace('\\', '')
        print(f"{name_clean:<25} {smooth_clean:<25} {rate_clean:<20}")
    print("-" * 80)


if __name__ == '__main__':
    main()

File: codes/julia/ch07/cheb_convergence.jl
# cheb_convergence.jl
#
# Demonstrates spectral convergence rates for four functions of increasing
# smoothness. Shows the fundamental relationship between function regularity
# and convergence rate of spectral differentiation.
#
# Test functions (different from Trefethen's standard examples):
# 1. |x|^(5/2) - Third derivative of bounded variation (algebraic: ~N^{-2.5})
# 2. exp(-1/(1-x^2)) - Smooth bump function, C^inf but not analytic (superalgebraic)
# 3. tanh(5x) - Analytic in [-1,1], poles at +-i*pi/10 (exponential)
# 4. x^8 - Polynomial of degree 8 (exact for N >= 8)
#
# This script generates Figure 6.5 and Table 6.2 for Chapter 6.
#
# Author: Dr. Denys Dutykh
#         Mathematics Department
#         Khalifa University of Science and Technology
#         Abu Dhabi, UAE
#
# Part of the book "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using Colors
using Printf

# Import Chebyshev matrix function
include(joinpath(@__DIR__, "cheb_matrix.jl"))

# -----------------------------------------------------------------------------
# Publication-quality CairoMakie configuration
# -----------------------------------------------------------------------------
set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 12,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 10,),
))

# -----------------------------------------------------------------------------
# Book color scheme
# -----------------------------------------------------------------------------
const NAVY   = colorant"#142D6E"
const SKY    = colorant"#7896D2"
const CORAL  = colorant"#E74C3C"
const TEAL   = colorant"#16A085"
const PURPLE = colorant"#9B59B6"
const ORANGE = colorant"#E67E22"

# Output paths
SCRIPT_DIR = @__DIR__
OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch07", "julia")
mkpath(OUTPUT_DIR)

# -----------------------------------------------------------------------------
# Test functions and their exact derivatives
# -----------------------------------------------------------------------------

# Function 1: |x|^(5/2) - C^2, 3rd derivative in BV
f1(x) = abs(x)^2.5
f1_prime(x) = 2.5 * abs(x)^1.5 * sign(x)

# Function 2: exp(-1/(1-x^2)) - Smooth bump, C^inf but not analytic at boundaries
function f2(x)
    abs(x) < 1.0 ? exp(-1.0 / (1.0 - x^2)) : 0.0
end

function f2_prime(x)
    if abs(x) < 1.0
        inner = 1.0 - x^2
        return exp(-1.0 / inner) * (-2.0 * x) / inner^2
    else
        return 0.0
    end
end

# Function 3: tanh(5x) - Analytic, poles at x = +-i*pi/10
f3(x) = tanh(5.0 * x)
f3_prime(x) = 5.0 / cosh(5.0 * x)^2

# Function 4: x^8 - Polynomial degree 8
f4(x) = x^8
f4_prime(x) = 8.0 * x^7

# List of test functions
const TEST_FUNCTIONS = [
    (name = L"|x|^{5/2}",         f = f1, fp = f1_prime,
     smoothness = L"C^2, 3rd deriv in BV", rate = L"O(N^{-2.5})",
     color = CORAL),
    (name = L"e^{-1/(1-x^2)}",    f = f2, fp = f2_prime,
     smoothness = L"C^\infty, not analytic", rate = L"faster than any $N^{-k}$",
     color = TEAL),
    (name = L"\tanh(5x)",          f = f3, fp = f3_prime,
     smoothness = "Analytic",       rate = L"O(\rho^{-N})",
     color = PURPLE),
    (name = L"x^8",                f = f4, fp = f4_prime,
     smoothness = "Polynomial",     rate = L"Exact for $N \geq 8$",
     color = ORANGE),
]

# -----------------------------------------------------------------------------
# Main visualization
# -----------------------------------------------------------------------------
function main()
    # Range of N values
    N_values = [4, 6, 8, 10, 12, 16, 20, 24, 32, 40, 48, 64, 80, 96, 128]

    # Create 2x2 figure
    fig = Figure(size = (700, 550))

    # Main title
    Label(fig[0, :], "Spectral Convergence: Four Functions of Increasing Smoothness",
          fontsize = 13, font = "CMU Serif Bold")

    positions = [(1, 1), (1, 2), (2, 1), (2, 2)]

    for (idx, (tf, pos)) in enumerate(zip(TEST_FUNCTIONS, positions))
        ax = Axis(fig[pos[1], pos[2]],
                  xlabel = L"N",
                  ylabel = "Max Error",
                  title = tf.name,
                  yscale = log10)
        hidespines!(ax, :t, :r)
        xlims!(ax, 0, 135)
        ylims!(ax, 1e-16, 1e2)

        # Grid
        ax.xgridvisible = true
        ax.xgridcolor = RGBAf(0.533, 0.533, 0.533, 0.3)
        ax.ygridvisible = true
        ax.ygridcolor = RGBAf(0.533, 0.533, 0.533, 0.3)

        errors = Float64[]

        for Nv in N_values
            D, x = cheb_matrix(Nv)
            v = tf.f.(x)
            w = D * v
            w_exact = tf.fp.(x)

            # Compute max error (avoiding boundary issues for bump function)
            if idx == 2  # Bump function
                interior = findall(xi -> xi > -0.99 && xi < 0.99, x)
                if !isempty(interior)
                    err = maximum(abs.(w[interior] .- w_exact[interior]))
                else
                    err = maximum(abs.(w .- w_exact))
                end
            else
                err = maximum(abs.(w .- w_exact))
            end

            push!(errors, max(err, 1e-16))
        end

        # Plot convergence
        scatterlines!(ax, Float64.(N_values), errors, color = tf.color,
                      linewidth = 1.5, markersize = 6,
                      strokecolor = :white, strokewidth = 0.5)

        # Add reference lines for theoretical rates
        if idx == 1  # |x|^(5/2) - algebraic O(N^-2.5)
            mask = N_values .>= 16
            N_ref = Float64.(N_values[mask])
            err_ref = errors[mask]
            if !isempty(err_ref)
                C = err_ref[1] * N_ref[1]^2.5
                lines!(ax, N_ref, C .* N_ref .^ (-2.5),
                       color = :gray, linewidth = 1, linestyle = :dash, alpha = 0.7,
                       label = L"O(N^{-2.5})")
            end
        elseif idx == 4  # x^8 - exact for N >= 8
            hlines!(ax, 1e-14, color = :gray, linewidth = 1, linestyle = :dash, alpha = 0.7)
        end

        # Add convergence rate annotation
        text!(ax, 125, 10^1.5,
              text = string(tf.rate),
              fontsize = 9, align = (:right, :top))
    end

    rowgap!(fig.layout, 1, 5)
    rowgap!(fig.layout, 2, 15)

    # Save figure
    save(joinpath(OUTPUT_DIR, "convergence_waterfall.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "convergence_waterfall.png"), fig, px_per_unit = 4)
    println("Figure saved to: $(joinpath(OUTPUT_DIR, "convergence_waterfall.pdf"))")

    # Generate convergence table
    println("\n" * "=" ^ 80)
    println("Convergence Table: Chebyshev Spectral Differentiation")
    println("=" ^ 80)

    # Table header
    @printf("%6s  %14s  %14s  %14s  %14s\n",
            "N", "|x|^(5/2)", "bump", "tanh(5x)", "x^8")
    println("-" ^ 80)

    # Table rows
    for Nv in N_values
        D, x = cheb_matrix(Nv)
        errs = Float64[]
        for (idx, tf) in enumerate(TEST_FUNCTIONS)
            v = tf.f.(x)
            w = D * v
            w_exact = tf.fp.(x)

            if idx == 2  # Bump function
                interior = findall(xi -> xi > -0.99 && xi < 0.99, x)
                if !isempty(interior)
                    err = maximum(abs.(w[interior] .- w_exact[interior]))
                else
                    err = maximum(abs.(w .- w_exact))
                end
            else
                err = maximum(abs.(w .- w_exact))
            end
            push!(errs, err)
        end

        @printf("%6d", Nv)
        for err in errs
            if err < 1e-14
                @printf("  %14s", "< 1e-14")
            else
                @printf("  %14.2e", err)
            end
        end
        println()
    end

    println("=" ^ 80)

    # Summary
    println("\nSummary:")
    println("-" ^ 80)
    @printf("%-25s %-25s %-20s\n", "Function", "Smoothness", "Expected Rate")
    println("-" ^ 80)
    summaries = [
        ("|x|^(5/2)",       "C^2, 3rd deriv in BV",    "O(N^(-2.5))"),
        ("exp(-1/(1-x^2))", "C^inf, not analytic",      "faster than any N^(-k)"),
        ("tanh(5x)",        "Analytic",                  "O(rho^(-N))"),
        ("x^8",             "Polynomial",                "Exact for N >= 8"),
    ]
    for (name, smooth, rate) in summaries
        @printf("%-25s %-25s %-20s\n", name, smooth, rate)
    end
    println("-" ^ 80)
end

main()
